Document Expected Features for the HTTP client integrations (#341)
getsentry/develop: src/docs/sdk/features.mdx

## HTTP Proxy

Ability to use an HTTP proxy. Often easy to implement using the existing HTTP client.  This should be picked up from the system config if possible or explicit config in the client options.

## HTTP Client Integrations

> Add breadcrumb for each outgoing HTTP request after the request finishes:

Reviewer: ```suggestion
Add a breadcrumb for each outgoing HTTP request after the request finishes:
```

- type: `http`
- category: `http`
- data: `url`, `method` (uppercase HTTP method), optionally `requestBodySize` and `responseBodySize` (in bytes)

> If the Performance Monitoring is supported by the SDK and enabled in the client application, when the transaction is active, a new `Span` must be created around the HTTP request.

Reviewer: ```suggestion
If Performance Monitoring is both supported by the SDK and enabled in the client application when the transaction is active a new `Span` must be created around the HTTP request.
```

> If Performance Monitoring is both supported by the SDK and enabled in the client application when the transaction is active a new `Span` must be created around the HTTP request.

Reviewer 2: describe span `op` and `desc`
eg `op=http.client` and `desc=$method $url`

also the OkHttp integration now has a `BeforeSpanCallback` which allows customizing the `span`

Reviewer 3: > describe span op and desc
> eg op=http.client and desc=$method $url

You mean just adding these examples? What description do you have in mind?

> also the OkHttp integration now has a BeforeSpanCallback which allows customizing the span

What's the suggestion? To add a note that integrations can take a callback to modify the span directly in a way that the HTTP Context is provided to the callback?

Reviewer 3: the PR looks good to me and got 👍 from @kamilogorek and other folks tagged had 1 month to react.
Happy to merge as-is or feel free to add the suggestions, commit them and merge.

🚀 🚀

Reviewer 2: my suggestion was just to explain that the `op` should be `http.client` and the description of the span should follow the convention `$method $url` eg `GET https://sentry.io`, so we don't diverge on that.

the `BeforeSpanCallback` can be added later since we're waiting feedback about it yet

Reviewer 2: ```suggestion
If Performance Monitoring is both supported by the SDK and enabled in the client application when the transaction is active a new `Span` must be created around the HTTP request; The `Span#operation` should be `http.client` and the `Span#description` should follow the convention `$METHOD $url` eg `GET https://sentry.io`.
```

> If Performance Monitoring is both supported by the SDK and enabled in the client application when the transaction is active a new `Span` must be created around the HTTP request; The `Span#operation` should be `http.client` and the `Span#description` should follow the convention `$METHOD $url` eg `GET https://sentry.io`.

Reviewer 2: ```suggestion
If Performance Monitoring is both supported by the SDK and enabled in the client application when the transaction is active a new `Span` must be created around the HTTP request:

- operation: `http.client`
- description: `$METHOD $url` (uppercase HTTP method), e.g. `GET https://sentry.io`
```

- HTTP request must be enhanced with a [`sentry-trace` HTTP header](/sdk/performance/#header-sentry-trace)
- span status must match HTTP response status code ([see Span status to HTTP status code mapping](/sdk/event-payloads/span/))

> - when network error occurs, span status must be set to `INTERNAL_SERVER_ERROR`

Reviewer 4: What is meant by "network error" here? There are many things that can happen like failed to resolve DNS, timeout, etc.

Note that `INTERNAL_SERVER_ERROR` is not a valid status (`internal_error` is).

Lacking more context, I guess `unknown_error` is more appropriate as a generic catch-all status, but I look forward to us simplifying the statuses in the near future, the current list is confusing and not well fitting the product.

Reviewer 2: @rhcarvalho about the `unknown_error `, See -> https://github.com/getsentry/develop/issues/307#issuecomment-822240763
we've changed it because the description of `unknown_error` is actually:

> An unknown error raised by APIs that don't return enough error information

and in our case, network error is eg you've no connection, something happened that you could not really talk with the endpoint, that's why internal rather unknown.

Reviewer 2: I agree that status are a bit confusing, but that's what we've right now, we can revisit this in the future

Reviewer 4: I see. Reusing `internal_error` unfortunately confuses with a 500 response.

Reviewer 2: indeed, but `unknown_error` is also 500 anyway, so it's not any better IMO

Reviewer 4: To be sure we understand the same, while the status values (first column, `States`) is sort of fixed ([known by Relay](https://github.com/getsentry/relay/blob/55127c75d4eeebf787848a05a12150ee5c59acd9/relay-common/src/constants.rs#L283-L306)), the `Description` and `HTTP status code equivalent` are under our control; it is whatever we want to document and it is relatively easy to change. AFAIK nothing in the product depends on those descriptions.

Reviewer 2: yep, but right now that's how its implemented in some of the SDKs (using inetrnal_error) and its the one that fits the most in the description https://develop.sentry.dev/sdk/event-payloads/transaction/
so unless we fix the SDKs and the status descriptions, let's keep it like this, I'll merge as it is.

> - when network error occurs, span status must be set to `INTERNAL_SERVER_ERROR`

Reviewer 2: ```suggestion
- when network error occurs, span status must be set to `internal_error`
```

## HTTP Proxy

Ability to use an HTTP proxy. Often easy to implement using the existing HTTP client.  This should be picked up from the system config if possible or explicit config in the client options.

## HTTP Client Integrations

Add a breadcrumb for each outgoing HTTP request after the request finishes:

- type: `http`
- category: `http`
- data: `url`, `method` (uppercase HTTP method), optionally `requestBodySize` and `responseBodySize` (in bytes)

If Performance Monitoring is both supported by the SDK and enabled in the client application when the transaction is active a new `Span` must be created around the HTTP request:

- operation: `http.client`
- description: `$METHOD $url` (uppercase HTTP method), e.g. `GET https://sentry.io`

> 

Reviewer 2: ```suggestion
```
